aws/resource_aws_budget_test.go

```diff
@@ -0,0 +1,333 @@
+package aws
+
+import (
+	"bytes"
+	"fmt"
+	"regexp"
+	"strconv"
+	"testing"
+	"text/template"
+	"time"
+
+	"github.com/aws/aws-sdk-go/service/budgets"
+	"github.com/hashicorp/terraform/helper/acctest"
+	"github.com/hashicorp/terraform/helper/resource"
+	"github.com/hashicorp/terraform/helper/schema"
+	"github.com/hashicorp/terraform/terraform"
+)
+
+func TestAwsBudget_basic(t *testing.T) {
+	name := fmt.Sprintf("test-budget-%d", acctest.RandInt())
+	configBasicDefaults := newBudgetTestConfigDefaults(name)
+	configBasicUpdate := newBudgetTestConfigUpdate(name)
+
+	resource.Test(t, resource.TestCase{
+		PreCheck:  func() { testAccPreCheck(t) },
+		Providers: testAccProviders,
+		CheckDestroy: func(s *terraform.State) error {
+			return testCheckBudgetDestroy(name, testAccProvider)
+		},
+		Steps: []resource.TestStep{
+			{
+				Config: testBudgetHCLBasicUseDefaults(configBasicDefaults),
+				Check:  newComposedBudgetTestCheck(configBasicDefaults, testAccProvider, "name"),
+			},
+
+			{
+				Config: testBudgetHCLBasic(configBasicUpdate),
+				Check:  newComposedBudgetTestCheck(configBasicUpdate, testAccProvider, "name"),
+			},
+		},
+	})
+}
+
+func TestAwsBudget_prefix(t *testing.T) {
+	name := "test-budget-"
+	configBasicDefaults := newBudgetTestConfigDefaults(name)
+	configBasicUpdate := newBudgetTestConfigUpdate(name)
+
+	resource.Test(t, resource.TestCase{
+		PreCheck:  func() { testAccPreCheck(t) },
+		Providers: testAccProviders,
+		CheckDestroy: func(s *terraform.State) error {
+			return testCheckBudgetDestroy(name, testAccProvider)
+		},
+		Steps: []resource.TestStep{
+			{
+				Config: testBudgetHCLPrefixUseDefaults(configBasicDefaults),
+				Check:  newComposedBudgetTestCheck(configBasicDefaults, testAccProvider, "name_prefix"),
+			},
+
+			{
+				Config: testBudgetHCLPrefix(configBasicUpdate),
+				Check:  newComposedBudgetTestCheck(configBasicUpdate, testAccProvider, "name_prefix"),
+			},
+		},
+	})
+}
+
+func newComposedBudgetTestCheck(config budgetTestConfig, provider *schema.Provider, nameField string) resource.TestCheckFunc {
```

Comment:
We do not need this extra function redirection (simpler is better when it comes to the testing).